### ISSUE TITLE
Many duplicate strings held by ProjectItemInstance.TaskItem._definingFileEscaped

### ISSUE THREAD

**Issue by drewnoakes** [microsoft/msbuild#4151]
Analysis of a devenv.exe dump shows the number one duplicated string on the heap (by wasted bytes) is held by `ProjectItemInstance.TaskItem._definingFileEscaped`. 

This field is populated by deserialisation/translation and not interned, at least in the code path used here.

Opening `Roslyn.sln` in 28606.22.d16.0stg, building and dumping the process reveals:

- `_definingFileEscaped` references 73,380 string objects (unique addresses), but only 238 unique string values
- The 73,380 strings occupy ~6.5MB bytes, 99.2% of which is wasted, meaning pooling here will reclaim ~6.5MB (0.6% of the ~1.2GB managed heap)

//cc: @davkean @lifengl

---

## Duplicated strings held by _definingFileEscaped

| WastedBytes | Count | Length | String |
|------------:|------:|-------:|--------|
| 4,881,282 | 16,604 | 140 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\Managed\Microsoft.Managed.DesignTime.targets |
| 1,015,774 | 3,878 | 124 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets |
| 199,436 | 684 | 139 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\Managed\Microsoft.CSharp.DesignTime.targets |
| 79,910 | 306 | 124 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets |
| 72,178 | 240 | 144 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\Managed\Microsoft.VisualBasic.DesignTime.targets |
| 62,328 | 197 | 152 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\v16.0\Fakes\Microsoft.QualityTools.Testing.Fakes.targets |
| 35,560 | 141 | 120 | D:\Program Files (x86)\Microsoft Visual Studio\2019\Preview2\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets |
| 19,716 | 63 | 152 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\v16.0\CodeSharing\Microsoft.SharedProject.Common.targets |
| 11,152 | 42 | 129 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Current\Bin\Microsoft.VisualBasic.CurrentVersion.targets |
| 10,912 | 45 | 117 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\NuGet\16.0\Microsoft.NuGet.targets |
| 5,916 | 18 | 167 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\Microsoft.NET.DefaultPackageConflictOverrides.targets |
| 4,082 | 14 | 150 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\v16.0\CodeSharing\Microsoft.CodeSharing.CSharp.targets |
| 2,548 | 27 | 42 | D:\repos\roslyn\eng\targets\Settings.props |
| 1,824 | 9 | 107 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets |
| 972 | 4 | 155 | D:\Program Files %28x86%29\Microsoft Visual Studio\2019\Preview2\MSBuild\Microsoft\VisualStudio\v16.0\CodeSharing\Microsoft.CodeSharing.VisualBasic.targets |
| 536 | 3 | 127 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets |
| 464 | 3 | 109 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\RepositoryInfo.targets |
| 460 | 3 | 108 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\ProjectDefaults.props |
| 432 | 3 | 101 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\Compiler.props |
| 298 | 2 | 142 | C:\Users\drnoakes\.nuget\packages\microsoft.visualstudio.threading.analyzers\15.8.168\build\Microsoft.VisualStudio.Threading.Analyzers.targets |
| 284 | 2 | 135 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.DefaultPackageConflictOverrides.targets |
| 278 | 2 | 132 | C:\Users\drnoakes\.nuget\packages\xunit.runner.visualstudio\2.4.1-pre.build.4059\build\netcoreapp1.0\xunit.runner.visualstudio.props |
| 268 | 2 | 127 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.SupportedTargetFrameworks.props |
| 264 | 2 | 125 | C:\Users\drnoakes\.nuget\packages\microsoft.sourcelink.vsts.git\1.0.0-beta-63401-01\build\Microsoft.SourceLink.Vsts.Git.props |
| 256 | 2 | 121 | C:\Users\drnoakes\.nuget\packages\microsoft.sourcelink.github\1.0.0-beta-63401-01\build\Microsoft.SourceLink.GitHub.props |
| 254 | 2 | 120 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.BeforeCommon.targets |
| 254 | 2 | 120 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets |
| 250 | 2 | 118 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.props |
| 248 | 2 | 117 | C:\Users\drnoakes\.nuget\packages\microsoft.net.test.sdk\15.9.0-dev2\build\netcoreapp1.0\Microsoft.NET.Test.Sdk.props |
| 244 | 2 | 115 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets |
| 232 | 2 | 109 | C:\Users\drnoakes\.nuget\packages\microsoft.netcore.app\2.1.0\build\netcoreapp2.1\Microsoft.NETCore.App.props |
| 228 | 2 | 107 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\Localization.targets |
| 226 | 2 | 106 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\XUnit\XUnit.targets |
| 214 | 2 | 100 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\Tests.targets |
| 214 | 2 | 100 | D:\repos\roslyn\src\Compilers\CSharp\Test\WinRT\Microsoft.CodeAnalysis.CSharp.WinRT.UnitTests.csproj |
| 210 | 2 | 98 | C:\Users\drnoakes\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19064.4\tools\Tests.props |
| 200 | 2 | 93 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Microsoft.NETCoreSdk.BundledVersions.props |
| 200 | 2 | 93 | C:\Program Files\dotnet\sdk\3.0.100-preview-009827\Microsoft.NETCoreSdk.BundledCliTools.props |
| 190 | 2 | 88 | C:\Users\drnoakes\.nuget\packages\xunit.core\2.4.1-pre.build.4059\build\xunit.core.props |
| 188 | 2 | 87 | C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WorkflowBuildExtensions.targets |
| 152 | 2 | 69 | D:\repos\roslyn\src\Compilers\Server\VBCSCompiler\VBCSCompiler.csproj |
| 150 | 2 | 68 | C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.Xaml.targets |
| 106 | 2 | 46 | D:\repos\roslyn\src\Interactive\csi\csi.csproj |
| 106 | 2 | 46 | D:\repos\roslyn\src\Interactive\vbi\vbi.vbproj |
| 100 | 2 | 43 | D:\repos\roslyn\eng\targets\Imports.targets |
| 82 | 2 | 34 | D:\repos\roslyn\eng\Versions.props |

**Comment by davkean** [comment]
I have a dump of a solution containing all of http://github.com/dotnet/corefx, and above is 2.5% of the heap:

![image](https://user-images.githubusercontent.com/1103906/52391563-d75a1200-2af1-11e9-95df-c37667adc151.png)

**Comment by davkean** [comment]
Note that _includeEscaped & _includeBeforeWildcardExpansionEscaped also suffer the same problem.

**Comment by davkean** [comment]
It looks like there is some interning going on for the metadata: https://github.com/Microsoft/msbuild/blob/eb3b0a79d52bf2029913ac50ad6fa64f7497d7f0/src/Build/Instance/ProjectItemInstance.cs#L1587-L1624.

Maybe just need to opt into this for other fields?

**Comment by davkean** [comment]
Though in saying that there's a path that doesn't get passed an interner where these fields are set:

https://github.com/Microsoft/msbuild/blob/eb3b0a79d52bf2029913ac50ad6fa64f7497d7f0/src/Build/Instance/ProjectItemInstance.cs#L1425-L1446

No idea when/when not either path are used.

**Comment by drewnoakes** [comment]
This analysis performed using [StringTheory](https://github.com/drewnoakes/string-theory). It may be useful in tracking down further interning/pooling opportunities.

**Comment by rainersigwald** [comment]
The lookaside interner doesn't help too much here: it only single-instances strings _within a single item_, so for things that are defined in a `.targets` file imported everywhere it hardly helps at all.

But I have a theory. Assigning this to me.

**Comment by panopticoncentral** [comment]
I took a deeper look at this, and I think that some of the analysis above is incorrect. The issue here doesn't seem to be related to serialization/deserialization and interning, that's not the codepath that's causing the vast majority of the wasted strings.

The telltale sign is the fact that about 97% of the wasted strings start with `D:\Program Files %28x86%29`. The codepath that's creating those strings is:

```
Microsoft.Build.Execution.ProjectItemInstance.TaskItem.TaskItem()
Microsoft.Build.dll!Microsoft.Build.Execution.ProjectItemInstance.CommonConstructor()
Microsoft.Build.dll!Microsoft.Build.Execution.ProjectInstance.CreateItemsSnapshot()
Microsoft.Build.dll!Microsoft.Build.Execution.ProjectInstance.ProjectInstance()
Microsoft.Build.dll!Microsoft.Build.Evaluation.Project.CreateProjectInstance()
Microsoft.VisualStudio.ProjectSystem.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.Designers.ProjectSnapshotService.GenerateProjectInstanceAsync()
```

So the snapshot service is creating a project instance and we're cloning the items. The offending routine is `CreateItemsSnapshot`. The line is:

`ProjectItemInstance instance = new ProjectItemInstance(this, item.ItemType, ((IItem)item).EvaluatedIncludeEscaped, item.EvaluatedIncludeBeforeWildcardExpansionEscaped, directMetadata, inheritedItemDefinitions, ProjectCollection.Escape(item.Xml.ContainingProject.FullPath));`

When we copy the item _for each item_ we are escaping the item's XML's containing project's full path, which creates a fresh string for each item if it contains unescaped characters. A simple fix would simply to have item.Xml.ContainingProject expose an EscapedFullPath property which would mean that the items would all share the same escaped path. I've tested this out and it seems like it fixes it. I'm out next week but I can prep a PR when I get back.
